torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -519,13 +543,58 @@ void XLAGraphExecutor::TensorCollectionBarrier(SyncTensorCollection* coll) {
 
 std::vector<torch::lazy::BackendDataPtr>
 XLAGraphExecutor::ExecuteComputationWithBarrier(
-    torch::lazy::ComputationPtr computation,
-    c10::ArrayRef<torch::lazy::BackendDataPtr> arguments,
+    torch::lazy::hash_t hash,
+    std::vector<torch::lazy::BackendDataPtr> arguments,
     const torch::lazy::BackendDevice& device) {
-  std::vector<torch::lazy::ExceptionCleanup> unlocker;
-  unlocker = DeviceLockerArena::Get()->LockDevices({device});
-  return torch::lazy::getBackend()->ExecuteComputation(computation, arguments,
-                                                       device);
+  MaybeDumpGraph("dynamo", hash);
+  auto cachedComputation =
+      XLAGraphExecutor::Get()->GetComputationCache()->Get(hash);
+  // TODO implement a fallback mechanism, or make sure those entries
+  // never get kicked out
+  XLA_CHECK(cachedComputation)
+      << "Failed to get computation by hash " << torch::lazy::HashToString(hash)
+      << ". Maybe the entry get "
+         "kicked out of the LRU cache";
+
+  // Create DataPlaceHolder that will get filled in async executions.
+  std::vector<xla::Shape>* output_shapes =
+      DeviceContextArena::Get()->GetOutputShapesByHash(hash);
+  std::vector<torch::lazy::BackendDataPtr> placeholders;
+  placeholders.reserve(output_shapes->size());
+  for (const xla::Shape& shape : *output_shapes) {
+    torch::lazy::BackendDataPtr handle =
+        WrapXlaData(xla::ComputationClient::Get()->CreateDataPlaceholder(
+            device.toString(), std::move(shape)));
+    placeholders.push_back(handle);
+  }
+
+  SyncTensorCollection coll;
+  coll.device = device;
+  coll.unlocker = DeviceLockerArena::Get()->LockDevices({device});
+  std::shared_ptr<XLAGraphExecutor::Async> async = std::make_shared<Async>(
+      &coll, std::move(arguments), placeholders, std::move(cachedComputation));
+
+  auto syncfn = [async, hash]() {
+    TF_VLOG(3) << "Executing Dynamo IR graph hash "
+               << torch::lazy::HashToString(hash) << " on device "
+               << async->device << " ...";
+    std::vector<torch::lazy::BackendDataPtr> results =
+        torch::lazy::getBackend()->ExecuteComputation(
+            async->cached_computation->computation, async->parameters_data,
+            async->device);
+    TF_VLOG(3) << "Executing Dynamo IR graph hash "
+               << torch::lazy::HashToString(hash) << " on device "
+               << async->device << " done!";
+    // Updating placeholder with actual output handle.
+    for (size_t i = 0; i < results.size(); ++i) {
+      XLA_CHECK(async->tensors_data[i] != nullptr);
+      async->tensors_data[i]->Assign(*results[i]);
+    }
+  };
+
+  xla::env::ScheduleIoClosure(async->mwait.Completer(std::move(syncfn)));
```

Comment:
sure, so if you look at above code
```
coll.unlocker = DeviceLockerArena::Get()->LockDevices({device});
std::shared_ptr<XLAGraphExecutor::Async> async = std::make_shared<Async>(
      &coll, std::move(arguments), placeholders, std::move(cachedComputation));
```

we grab the device lock and pass it to the `syncfn`. Every operation that requires reads real data from the device requires a device lock. Async execution thread will hold the device lock until it finished computation, and release the lock when stack object is destroyed. If user wants to access tensor's data like `t1.cpu()`, it will try to grab the lock which will force them to wait for async execution to finish.

Reply:
Thanks for the explanation!

One more thing confirm, so the lock is not a tensor level lock but a device level lock?  i.e., for each device, there is a single such lock?

Reply:
yes, it is a device level lock, hence the name `LockDevices`